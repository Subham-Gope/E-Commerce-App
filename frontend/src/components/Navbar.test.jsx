import { describe, test, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import ShopContext from "../context/ShopContext";
import Navbar from "./Navbar";

// Mocking the assests
vi.mock("../assets/assets", () => ({
  assets: {
    logo: "logo.png",
    search_icon: "search.png",
    profile_icon: "profile.png",
    cart_icon: "cart.png",
    menu_icon: "menu.png",
    dropdown_icon: "dropdown.png",
  },
}));

// default mock values for context
const defaultMockContext = {
  setShowSearch: vi.fn(),
  getCartCount: () => 0,
  navigate: vi.fn(),
  token: "",
  setToken: vi.fn(),
  setCartItems: vi.fn(),
};

// render with context
const renderWithContext = (ui, customValues = {}) => {
  const contextValue = { ...defaultMockContext, ...customValues };

  return render(
    <MemoryRouter>
      <ShopContext.Provider value={contextValue}>{ui}</ShopContext.Provider>
    </MemoryRouter>
  );
};

describe("Testing the Navbar.jsx file", () => {
  test("renders the navigation menu", async () => {
    // setting up the userEvent
    const user = userEvent.setup({ delay: 1 });

    // rendering and setting up the snapshot matching
    const { container } = renderWithContext(<Navbar />, {});
    expect(container).toMatchSnapshot();

    // checking if logo is rendered
    const logo = screen.getByAltText(/logo/i);
    expect(logo).toBeInTheDocument();

    // checking  home link is rendered
    const homeText = screen.getByText("HOME");
    expect(homeText).toHaveTextContent("HOME");

    // checking the search icon and setShowSearch function
    const searchIcon = screen.getByAltText(/search icon/i);
    expect(searchIcon).toBeInTheDocument();

    // Initiating user interaction
    await user.click(searchIcon);
    expect(defaultMockContext.setShowSearch).toHaveBeenCalled();

    // checking the profile icon and navigate function
    const profileIcon = screen.getByAltText(/profile icon/i);
    expect(profileIcon).toBeInTheDocument();

    await user.click(profileIcon);
    expect(defaultMockContext.navigate).toHaveBeenCalled();

    // checking the menu Icon
    const menuIcon = screen.getByAltText(/menu icon/i);
    expect(menuIcon).toBeInTheDocument();

    // selecting the the responsive sidebar
    const sidebar = screen.getByRole("sidebar");
    expect(sidebar).toBeInTheDocument();

    await user.click(menuIcon);

    expect(sidebar).toHaveClass("w-full");
    const backIcon = screen.getByAltText(/dropdown icon/i);
    expect(backIcon).toBeInTheDocument();

    await user.click(backIcon);

    expect(sidebar).toHaveClass("w-0");
  });
});
