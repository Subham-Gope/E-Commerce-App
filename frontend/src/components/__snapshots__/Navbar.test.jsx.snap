// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the Navbar.jsx file > renders the navigation menu 1`] = `
<div>
  <div
    class="flex items-center justify-between py-5 font-medium"
  >
    <a
      data-discover="true"
      href="/"
    >
      <img
        alt="Logo"
        src="logo.png"
      />
    </a>
    <ul
      class="hidden sm:flex gap-5 text-sm text-gray-700"
    >
      <a
        aria-current="page"
        class="flex flex-col items-center gap-1 active"
        data-discover="true"
        href="/"
      >
        <p>
          HOME
        </p>
        <hr
          class="w-2/4 border-none h-[1.5px] bg-gray-700 hidden"
        />
      </a>
      <a
        class="flex flex-col items-center gap-1"
        data-discover="true"
        href="/collection"
      >
        <p>
          COLLECTION
        </p>
        <hr
          class="w-2/4 border-none h-[1.5px] bg-gray-700 hidden"
        />
      </a>
      <a
        class="flex flex-col items-center gap-1"
        data-discover="true"
        href="/about"
      >
        <p>
          ABOUT
        </p>
        <hr
          class="w-2/4 border-none h-[1.5px] bg-gray-700 hidden"
        />
      </a>
      <a
        class="flex flex-col items-center gap-1"
        data-discover="true"
        href="/contact"
      >
        <p>
          CONTACT
        </p>
        <hr
          class="w-2/4 border-none h-[1.5px] bg-gray-700 hidden"
        />
      </a>
    </ul>
    <div
      class="flex items-center gap-6"
    >
      <img
        alt="Search Icon"
        class="w-5 cursor-pointer"
        src="search.png"
      />
      <div
        class="group realtive"
      >
        <img
          alt="Profile Icon"
          class="w-5 cursor-pointer"
          src="profile.png"
        />
      </div>
      <a
        class="relative"
        data-discover="true"
        href="/cart"
      >
        <img
          alt="cart icon"
          class="w-5 min-w-5"
          src="cart.png"
        />
        <p
          class="absolute right-[-5px] bottom-[-5px] w-4 text-center leading-4 bg-black text-white aspect-square rounded-full text-[8px]"
        >
          0
        </p>
      </a>
      <img
        alt="menu icon"
        class="w-5 cursor-pointer sm:hidden"
        src="menu.png"
      />
    </div>
    <div
      class="absolute top-0 right-0 bottom-0 overflow-hidden bg-white transition-all w-0"
      role="sidebar"
    >
      <div
        class="flex flex-col text-gray-600 gap-43"
      >
        <div
          class="flex items-center gap-4 p-3 w-fit cursor-pointer"
        >
          <img
            alt="dropdown icon"
            class="h-4 rotate-180"
            src="dropdown.png"
          />
          <p>
            Back
          </p>
        </div>
        <a
          aria-current="page"
          class="py-2 pl-6 border-none active"
          data-discover="true"
          href="/"
        >
          Home
        </a>
        <a
          class="py-2 pl-6 border-none"
          data-discover="true"
          href="/collection"
        >
          Collection
        </a>
        <a
          class="py-2 pl-6 border-none"
          data-discover="true"
          href="/about"
        >
          About
        </a>
        <a
          class="py-2 pl-6 border-none"
          data-discover="true"
          href="/contact"
        >
          Contact
        </a>
      </div>
    </div>
  </div>
</div>
`;
